cmd/mirroring/main.go

```diff
@@ -54,17 +55,56 @@ func readLogInfo(treeSize *int64, root *string) error {
 		line = scanner.Text()
 	}
 
-	// Each line is in the format of space-separeted info: "treeSize rootHash"
-	parsed := strings.Split(line, " ")
-	*treeSize, err = strconv.ParseInt(parsed[0], 10, 64)
+	sth := util.SignedCheckpoint{}
+	if err := sth.UnmarshalText([]byte(strings.Replace(line, "\\n", "\n", -1))); err != nil {
+		return nil, err
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return &sth, nil
+}
+
+// deleteOldCheckpoints persists the last 100 checkpoints in the log file
+func deleteOldCheckpoints(logInfoFile string) error {
+	// read all lines from file
+	file, err := os.Open(logInfoFile)
 	if err != nil {
 		return err
 	}
-	*root = parsed[1]
-
+	scanner := bufio.NewScanner(file)
+	var lines []string
+	for scanner.Scan() {
+		lines = append(lines, scanner.Text())
+	}
 	if err := scanner.Err(); err != nil {
 		return err
 	}
+	if err := file.Close(); err != nil {
+		return err
+	}
+
+	// exit early if there aren't checkpoints to truncate
+	if len(lines) < 100 {
+		return nil
+	}
+
+	// open file again to overwrite
+	file, err = os.OpenFile(logInfoFile, os.O_RDWR|os.O_TRUNC, 0666)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	// keep all but the oldest checkpoint
```

Comment:
Instead of a scanner, use a `bufio.NewReader` to `ReadBytes` for 100 lines, and maintain the end position, then just truncate the file to the calculated length.

something like
```
r := bufio.NewReader(file)
endPos := 0
for numLines := 0;  numLines < 100; numLines++ {
    line, err = r.ReadBytes('\n')
    if err != nil { return err }
    endPos += len(line)
}
if err := os.Truncate(logInfoFile, endPos); err != nil {
   return err
}
```

Reply:
The problem is we need to drop the first X lines from the file, not the last. Truncate drops only the end, instead of the beginning.

Reply:
That's even easier! But that's not what this function is doing. The latest checkpoint is at the start, right? Trunacting would drop the oldest ones.

Reply:
I started rewriting this but realized we still need a scanner to confirm how many lines are in the file. You could do something like counting `\n`, but I don't see much benefit to that. Also planning to remove this code once we start writing history to git.

Reply:
I am updating this to correctly persist only the last 100, not just the single oldest.